plottr/plot/base.py

```diff
@@ -134,3 +144,315 @@ def makeFlowchartWithPlot(nodes: List[Tuple[str, Type[Node]]],
+        if len(data.axes()) == 2:
+            return PlotDataType.scatter2d
+        else:
+            return PlotDataType.scatter1d
+
+    return PlotDataType.unknown
+
+
+@dataclass
+class PlotItem:
+    data: List[np.ndarray]
+    id: int
+    subPlot: int
+    labels: Optional[List[str]] = None
+    style: Optional[str] = None
+    plotOptions: Optional[Dict[str, Any]] = None
+    plotReturn: Optional[Any] = None
+
+
+@dataclass
+class SubPlot:
+    id: int
+    axes: Optional[List[Any]] = None
+
+
+class AutoFigureMaker(object):
+    """A class for semi-automatic creation of plot figures.
+    It must be inherited to tie it to a specific plotting backend.
+
+    The main purpose of this class is to (a) implement actual plotting of
+    plot items, and (b) distribute plot items correctly among subpanels of
+    a figure.
+
+    FigureMaker is a context manager. The user should eventually only need to
+    add data and specify what kind of data it is. FigureMaker will then
+    generate plots from that.
+
+    In the simplest form, usage looks something like this::
+
+        >>> with FigureMaker() as fm:
+        >>>     fm.addData(x, y, [...])
+        >>>     [...]
+
+    See :method:`addData` for details on how to specify data and how to pass
+    plot options to it.
+    """
+
+    # TODO: implement feature for always plotting certain traces with other
+    #   other ones ('children'). This is mainly used for models/fits.
+    #   needs a system to copy certain style aspects from the parents.
+    # TODO: similar, but with siblings (imagine Re/Im parts)
+
+    def __init__(self):
+        self.subPlots: OrderedDictType[int, SubPlot] = OrderedDict()
+        self.plotItems: OrderedDictType[int, PlotItem] = OrderedDict()
+
+        #: how to represent complex data.
+        #: must be set before adding data to the plot to have an effect.
+        self.complex_representation = ComplexRepresentation.realAndImag
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        self._makeAxes()
+        for id in self.subPlots.keys():
+            self._makeSubPlot(id)
+
+    # private methods
+    def _makeAxes(self) -> None:
+        n = self.nSubPlots()
+        for id, axes in zip(range(n), self.makeAxes(n)):
+            if not isinstance(axes, list):
+                axes = [axes]
+            self.subPlots[id] = SubPlot(id, axes)
+
+    def addPanel(self) -> int:
+        id = _generate_auto_dict_key(self.subPlots)
+        self.subPlots[id] = SubPlot(id, [])
+        return id
+
+    def _makeSubPlot(self, id: int):
+        items = self.subPlotItems(id)
+        for name, item in items.items():
+            self._plot(item)
+        self.formatSubPlot(id)
+
+    def _plot(self, plotItem: PlotItem):
+        if plotItem.style is None:
+            if len(plotItem.data) == 2:
+                style = 'line'
+            elif len(plotItem.data) == 3:
+                style = 'image'
```

Comment:
what's the difference between this "style" and "PlotType" enum? also, shouldn't determinePlotDataType be used here somehwere?

Reply:
As I understand it style is basically:

```1D or 2D```

```PlotDataType``` and ```PlotType``` are further refinements of this. 

Perhaps we should consider if we can streamline these types. My understanding is basically that we can categorize the data as follows. I like to think of it as a mapping from data types to valid plot types

* 1d data ordered: [1D line plot, 1D points plot (scatter)]
* 1d data not ordered: [1D points plot (scatter)]
* 2d data on equidistant grid: [image, colormesh, scatter2d]
* 2d data on non equidistant grid: [colormesh, scatter2d]
* 2d data no structure: [scatter2d]

Further to that there is choice between single and multitrace 1d plots which seems to be somewhat unrelated.

it seems to me that these types currently mix things that should be separate concerns but since that code is the same as it was before in mpl.py we should probably not touch that in this pr

Reply:
good point, Mikhail. this could essentially be the same thing. the idea is that any backend could essentially support different types of plots that for instance also the user may want to implement (i'm thinking of a few custom plots that would be useful for specific measurements).  Other backends might support 3D graphics in a decent way. etc...

Jens, agreed. The only thing i'd want to make sure of is that adding plot types should be somewhat easy, and potentially doable through a plugin rather than only by modifying 'core' code. 

I'm definitely open to come up with a structure that does things 'right' in this effort, in terms of separation. There's no immediate rush with merging this into the main branch.